File: backend/src/main/kotlin/app/meals/storage/RecipeRepository.kt
package app.meals.storage

import app.meals.domain.RecipeDoc
import java.sql.Connection
import java.sql.Types
import java.util.UUID

object RecipeRepository {
    fun count(): Long {
        return DatabaseFactory.query { conn ->
            conn.prepareStatement("SELECT COUNT(*) FROM recipes").use { ps ->
                ps.executeQuery().use { rs ->
                    rs.next()
                    rs.getLong(1)
                }
            }
        }
    }

    fun findAll(familyId: UUID, nameQuery: String? = null, tagQuery: String? = null): List<Pair<UUID, RecipeDoc>> {
        return DatabaseFactory.query { conn ->
            val sql = buildString {
                append("SELECT id, doc FROM recipes WHERE family_id = ? ")
                if (!nameQuery.isNullOrBlank()) append("AND doc->>'name' ILIKE ? ")
                // Use @> jsonb_build_array(?) instead of doc->'tags' ? tag: JDBC treats "?" as bind placeholders,
                // which breaks the jsonb "contains key/element" operator in prepared statements.
                if (!tagQuery.isNullOrBlank()) append("AND doc->'tags' @> jsonb_build_array(?) ")
                append("ORDER BY doc->>'name'")
            }
            conn.prepareStatement(sql).use { ps ->
                var i = 1
                ps.setObject(i++, familyId)
                if (!nameQuery.isNullOrBlank()) ps.setString(i++, "%$nameQuery%")
                if (!tagQuery.isNullOrBlank()) ps.setString(i, tagQuery)
                ps.executeQuery().use { rs ->
                    buildList {
                        while (rs.next()) {
                            val id = UUID.fromString(rs.getString("id"))
                            val docStr = rs.getJsonb("doc") ?: continue
                            add(id to docStr.fromJsonb<RecipeDoc>())
                        }
                    }
                }
            }
        }
    }

    fun findById(familyId: UUID, id: UUID): RecipeDoc? {
        return DatabaseFactory.query { conn ->
            conn.prepareStatement("SELECT doc FROM recipes WHERE id = ? AND family_id = ?").use { ps ->
                ps.setObject(1, id)
                ps.setObject(2, familyId)
                ps.executeQuery().use { rs ->
                    if (rs.next()) rs.getJsonb("doc")?.fromJsonb() else null
                }
            }
        }
    }

    fun insert(familyId: UUID, doc: RecipeDoc): UUID {
        return DatabaseFactory.query { conn ->
            insert(conn, familyId, doc)
        }
    }

    fun insert(conn: Connection, familyId: UUID, doc: RecipeDoc): UUID {
        return conn.prepareStatement("INSERT INTO recipes (family_id, doc) VALUES (?, ?::jsonb) RETURNING id").use { ps ->
            ps.setObject(1, familyId)
            ps.setJsonb(2, doc.toJsonbString())
            ps.executeQuery().use { rs ->
                check(rs.next()) { "INSERT RETURNING id must return one row" }
                UUID.fromString(rs.getString("id"))
            }
        }
    }

    fun update(familyId: UUID, id: UUID, doc: RecipeDoc): Boolean {
        return DatabaseFactory.query { conn ->
            conn.prepareStatement(
                "UPDATE recipes SET doc = ?::jsonb, updated_at = now() WHERE id = ? AND family_id = ?"
            ).use { ps ->
                ps.setJsonb(1, doc.toJsonbString())
                ps.setObject(2, id)
                ps.setObject(3, familyId)
                ps.executeUpdate() > 0
            }
        }
    }

    fun delete(familyId: UUID, id: UUID): Boolean {
        return DatabaseFactory.query { conn ->
            conn.prepareStatement("DELETE FROM recipes WHERE id = ? AND family_id = ?").use { ps ->
                ps.setObject(1, id)
                ps.setObject(2, familyId)
                ps.executeUpdate() > 0
            }
        }
    }

    fun suggestTags(familyId: UUID, prefix: String, excludeRecipeId: UUID?, limit: Int = 20): List<String> {
        val trimmed = prefix.trim()
        if (trimmed.isEmpty()) return emptyList()
        val pattern = "$trimmed%"
        return DatabaseFactory.query { conn ->
            val sql =
                """
                SELECT DISTINCT t.tag AS tag
                FROM recipes r
                CROSS JOIN LATERAL jsonb_array_elements_text(r.doc->'tags') AS t(tag)
                WHERE r.family_id = ?
                  AND (?::uuid IS NULL OR r.id <> ?::uuid)
                  AND t.tag ILIKE ?
                ORDER BY t.tag
                LIMIT ?
                """.trimIndent()
            conn.prepareStatement(sql).use { ps ->
                ps.setObject(1, familyId)
                if (excludeRecipeId == null) {
                    ps.setNull(2, Types.OTHER)
                    ps.setNull(3, Types.OTHER)
                } else {
                    ps.setObject(2, excludeRecipeId)
                    ps.setObject(3, excludeRecipeId)
                }
                ps.setString(4, pattern)
                ps.setInt(5, limit)
                ps.executeQuery().use { rs ->
                    buildList {
                        while (rs.next()) {
                            add(rs.getString("tag"))
                        }
                    }
                }
            }
        }
    }

    fun findByIds(familyId: UUID, ids: List<UUID>): List<Pair<UUID, RecipeDoc>> {
        if (ids.isEmpty()) return emptyList()
        return DatabaseFactory.query { conn ->
            val placeholders = ids.joinToString(",") { "?" }
            conn.prepareStatement(
                "SELECT id, doc FROM recipes WHERE family_id = ? AND id IN ($placeholders)"
            ).use { ps ->
                var i = 1
                ps.setObject(i++, familyId)
                ids.forEach { id -> ps.setObject(i++, id) }
                ps.executeQuery().use { rs ->
                    buildList {
                        while (rs.next()) {
                            val id = UUID.fromString(rs.getString("id"))
                            val docStr = rs.getJsonb("doc") ?: continue
                            add(id to docStr.fromJsonb<RecipeDoc>())
                        }
                    }
                }
            }
        }
    }

    fun moveAllToFamily(conn: Connection, fromFamilyId: UUID, toFamilyId: UUID): Int {
        conn.prepareStatement("UPDATE recipes SET family_id = ? WHERE family_id = ?").use { ps ->
            ps.setObject(1, toFamilyId)
            ps.setObject(2, fromFamilyId)
            return ps.executeUpdate()
        }
    }
}

File: backend/src/storage/recipeRepository.ts
import { sql } from "kysely";
import type { Db } from "../db/kysely.js";
import type { ConcurrentWriteResult, RecipeDoc, RecipeResponse } from "../domain/types.js";

export interface FindAllOptions {
  nameQuery?: string;
  tagQuery?: string;
}

export class RecipeRepository {
  constructor(private readonly db: Db) {}

  async count(): Promise<number> {
    const row = await this.db
      .selectFrom("recipes")
      .select((eb) => eb.fn.countAll<string>().as("count"))
      .executeTakeFirstOrThrow();
    return Number(row.count);
  }

  async findAll(familyId: string, options: FindAllOptions = {}): Promise<RecipeResponse[]> {
    let query = this.db
      .selectFrom("recipes")
      .select(["id", "doc", "version"])
      .where("family_id", "=", familyId);

    if (options.nameQuery?.trim()) {
      query = query.where(sql<boolean>`doc->>'name' ilike ${`%${options.nameQuery}%`}`);
    }
    if (options.tagQuery?.trim()) {
      // Cast to text: node-postgres binds params as `unknown`, which would make
      // jsonb_build_array's polymorphic type undeterminable.
      query = query.where(sql<boolean>`doc->'tags' @> jsonb_build_array(${options.tagQuery}::text)`);
    }

    const rows = await query.orderBy(sql`doc->>'name'`).execute();
    return rows.map((r) => ({ id: r.id, doc: r.doc, version: r.version }));
  }

  async findById(
    familyId: string,
    id: string,
  ): Promise<{ doc: RecipeDoc; version: number } | undefined> {
    const row = await this.db
      .selectFrom("recipes")
      .select(["doc", "version"])
      .where("id", "=", id)
      .where("family_id", "=", familyId)
      .executeTakeFirst();
    return row ? { doc: row.doc, version: row.version } : undefined;
  }

  async findByIds(familyId: string, ids: string[]): Promise<RecipeResponse[]> {
    if (ids.length === 0) return [];
    const rows = await this.db
      .selectFrom("recipes")
      .select(["id", "doc", "version"])
      .where("family_id", "=", familyId)
      .where("id", "in", ids)
      .execute();
    return rows.map((r) => ({ id: r.id, doc: r.doc, version: r.version }));
  }

  /**
   * Insert a recipe. When `id` is supplied (a client-minted UUID for
   * offline-first creates), it is used as the primary key and the insert is
   * idempotent: replaying the same create — e.g. after an offline outbox
   * survives a reload — conflicts on the id and does nothing, so the original
   * row wins. Without an `id` the database mints one as before.
   */
  async insert(
    familyId: string,
    doc: RecipeDoc,
    id?: string,
  ): Promise<{ id: string; version: number }> {
    if (id) {
      const inserted = await this.db
        .insertInto("recipes")
        .values({ id, family_id: familyId, doc: JSON.stringify(doc) })
        .onConflict((oc) => oc.column("id").doNothing())
        .returning(["id", "version"])
        .executeTakeFirst();
      if (inserted) return { id: inserted.id, version: inserted.version };
      // Idempotent replay: the row already existed, so return its current
      // version (the original doc wins — see the doc comment above).
      const existing = await this.findById(familyId, id);
      return { id, version: existing?.version ?? 0 };
    }
    const row = await this.db
      .insertInto("recipes")
      .values({ family_id: familyId, doc: JSON.stringify(doc) })
      .returning(["id", "version"])
      .executeTakeFirstOrThrow();
    return { id: row.id, version: row.version };
  }

  /**
   * Optimistic-concurrency update. Without a `baseVersion` the write is
   * unconditional (preserving legacy single-client behaviour); with one, the
   * update only lands when the stored version matches, bumping it on success.
   * A mismatch returns the current server doc + version so the caller can 409.
   */
  async update(
    familyId: string,
    id: string,
    doc: RecipeDoc,
    baseVersion?: number,
  ): Promise<ConcurrentWriteResult<RecipeDoc>> {
    let update = this.db
      .updateTable("recipes")
      .set({ doc: JSON.stringify(doc), updated_at: new Date(), version: sql`version + 1` })
      .where("id", "=", id)
      .where("family_id", "=", familyId);
    if (baseVersion !== undefined) {
      update = update.where("version", "=", baseVersion);
    }
    const updated = await update.returning("version").executeTakeFirst();
    if (updated) return { status: "updated", version: updated.version };

    // Nothing updated: distinguish a missing row from a version mismatch.
    const current = await this.findById(familyId, id);
    if (!current) return { status: "notFound" };
    return { status: "conflict", doc: current.doc, version: current.version };
  }

  async delete(familyId: string, id: string): Promise<boolean> {
    const result = await this.db
      .deleteFrom("recipes")
      .where("id", "=", id)
      .where("family_id", "=", familyId)
      .executeTakeFirstOrThrow();
    return result.numDeletedRows > 0n;
  }

  /** Distinct tags matching a prefix across the family's recipes (excluding one recipe). */
  async suggestTags(
    familyId: string,
    prefix: string,
    excludeRecipeId?: string,
    limit = 20,
  ): Promise<string[]> {
    const trimmed = prefix.trim();
    if (!trimmed) return [];
    const exclude = excludeRecipeId ?? null;
    const result = await sql<{ tag: string }>`
      SELECT DISTINCT t.tag AS tag
      FROM recipes r
      CROSS JOIN LATERAL jsonb_array_elements_text(r.doc->'tags') AS t(tag)
      WHERE r.family_id = ${familyId}
        AND (${exclude}::uuid IS NULL OR r.id <> ${exclude}::uuid)
        AND t.tag ILIKE ${`${trimmed}%`}
      ORDER BY t.tag
      LIMIT ${limit}
    `.execute(this.db);
    return result.rows.map((r) => r.tag);
  }
}
